src/Engine/Data/EnvironmentTexture.cpp

```diff
@@ -0,0 +1,739 @@
+        m_skyData[imgIdx] = new float[m_width * m_height * 4];
+        for ( size_t l = 0; l < m_height; ++l ) {
+            for ( size_t c = 0; c < m_width; ++c ) {
+                auto is                       = ( l * m_width + c );
+                auto id                       = flipV ? is : ( l * m_width + ( m_width - c - 1 ) );
+                m_skyData[imgIdx][id * 4 + 0] = loaded[is * n + 0];
+                m_skyData[imgIdx][id * 4 + 1] = loaded[is * n + 1];
+                m_skyData[imgIdx][id * 4 + 2] = loaded[is * n + 2];
+                m_skyData[imgIdx][id * 4 + 3] = 0;
+            }
+        }
+        stbi_image_free( loaded );
+    }
+}
+
+void EnvironmentTexture::setupTexturesFromSphericalEquiRectangular() {
+    auto ext = m_name.substr( m_name.size() - 3 );
+    float* latlonPix { nullptr };
+    int n, w, h;
+    if ( ext == "exr" ) {
+        const char* err { nullptr };
+        // EXR image are always in linear RGB color space
+        int ret = LoadEXR( &latlonPix, &w, &h, m_name.c_str(), &err );
+        n       = 4;
+        if ( ret != TINYEXR_SUCCESS ) {
+            if ( err ) {
+                std::cerr << "Error reading " << m_name << " : " << err << std::endl;
+                FreeEXRErrorMessage( err ); // release memory of error message.
+            }
+        }
+    }
+    else {
+        stbi_set_flip_vertically_on_load( false );
+        // Assume input images are in sRGB color space. stbi_loadf will convert to linear space.
+        latlonPix = stbi_loadf( m_name.c_str(), &w, &h, &n, 4 );
+    }
+    int textureSize = 1;
+    while ( textureSize < h ) {
+        textureSize <<= 1;
+    }
+    textureSize >>= 1;
+    // Bases to use to convert sphericalequirectangular images to cube faces
+    // These bases allow to convert (u, v) cordinates of each faces to (x, y, z) in the frame of the
+    // equirectangular map. : (x, y, z) = u*A[0] + v*A[1] + A[2]
+    Vector3 bases[6][3] = { { { -1, 0, 0 }, { 0, 1, 0 }, { 0, 0, -1 } },
+                            { { 1, 0, 0 }, { 0, -1, 0 }, { 0, 0, -1 } },
+                            { { 0, 0, 1 }, { 1, 0, 0 }, { 0, 1, 0 } },
+                            { { 0, 0, -1 }, { 1, 0, 0 }, { 0, -1, 0 } },
+                            { { 0, 1, 0 }, { 1, 0, 0 }, { 0, 0, -1 } },
+                            { { 0, -1, 0 }, { -1, 0, 0 }, { 0, 0, -1 } } };
+    auto sphericalPhi   = []( const Vector3& d ) {
+        Scalar p = std::atan2( d.x(), d.y() );
+        return ( p < 0 ) ? ( p + 2 * M_PI ) : p;
+    };
+    auto sphericalTheta = []( const Vector3& d ) { return std::acos( d.z() ); };
+#pragma omp parallel for
+    for ( int imgIdx = 0; imgIdx < 6; ++imgIdx ) {
+        // Alllocate the images
+        m_skyData[imgIdx] = new float[textureSize * textureSize * 4];
+        // Fill in pixels
+        for ( Scalar u = -1_ra; u < 1_ra; u += 2_ra / textureSize ) {
+            for ( Scalar v = -1_ra; v < 1_ra; v += 2_ra / textureSize ) {
```

Comment:
I didn't try, but we could go to something like that:

```suggestion
// Alllocate the images
for ( int imgIdx = 0; imgIdx < 6; ++imgIdx )
    m_skyData[imgIdx] = new float[textureSize * textureSize * 4];
    
// Fill in pixels
#pragma omp parallel for collapse (3)
    for ( int imgIdx = 0; imgIdx < 6; ++imgIdx ) {
        for ( Scalar u = -1_ra; u < 1_ra; u += 2_ra / textureSize ) {
            for ( Scalar v = -1_ra; v < 1_ra; v += 2_ra / textureSize ) {
```

Reply:
In my test, I do not see any improvement with collapse